### PR TITLE
feat: dynamic rate limits

### PR DESCRIPTION
**Pull request by abelanger5**
# Description

Adds support for rate limit units, keys, and values computed at runtime via CEL expressions. Docs and implementations for other SDKs to come.

Also adds a rate limit view to the UI.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: api-contracts/workflows/workflows.proto [hunk 1]
```diff
@@ -117,7 +117,11 @@ message CreateWorkflowStepOpts {
 
 message CreateStepRateLimit {
     string key = 1; // (required) the key for the rate limit
-    int32 units = 2; // (required) the number of units this step consumes
+    optional int32 units = 2; // (required) the number of units this step consumes
```
Comment (reillyse) [inline api-contracts/workflows/workflows.proto]: this comment probably needs updating ?

File: pkg/repository/prisma/workflow.go [hunk 2]
```diff
@@ -923,16 +923,82 @@ func (r *workflowEngineRepository) createJobTx(ctx context.Context, tx pgx.Tx, t
...
+			}
+
 			for _, rateLimit := range stepOpts.RateLimits {
-				_, err := r.queries.CreateStepRateLimit(
+				// if ANY of the step expressions are not nil, we create ALL options as expressions, but with static
+				// keys for any nil expressions.
+				if rateLimit.KeyExpr != nil || rateLimit.LimitExpr != nil || rateLimit.UnitsExpr != nil {
+					var keyExpr, limitExpr, unitsExpr string
+
+					// FIXME: move all these double quotes to a shared CEL method
```
Comment (reillyse) [inline pkg/repository/prisma/workflow.go]: fixme

File: pkg/repository/prisma/workflow.go [hunk 3]
```diff
@@ -923,16 +923,82 @@ func (r *workflowEngineRepository) createJobTx(ctx context.Context, tx pgx.Tx, t
...
+						createStepExprParams.Keys = append(createStepExprParams.Keys, rateLimit.Key)
+						createStepExprParams.Expressions = append(createStepExprParams.Expressions, limitExpr)
+					}
+
+					// create the units value expression
+					createStepExprParams.Kinds = append(createStepExprParams.Kinds, string(dbsqlc.StepExpressionKindDYNAMICRATELIMITUNITS))
+					createStepExprParams.Keys = append(createStepExprParams.Keys, rateLimit.Key)
+					createStepExprParams.Expressions = append(createStepExprParams.Expressions, unitsExpr)
+
+					// create the durations expression (dunno why we're calling it window in some places)
```
Comment (reillyse) [inline pkg/repository/prisma/workflow.go]: Maybe correct refs to duration so it's consistent.

**Review by reillyse — APPROVED** [review]
couple of nits, LGTM

File: examples/rate-limit/main.go [hunk 4]
```diff
@@ -76,8 +81,10 @@ func main() {
 			Steps: []*worker.WorkflowStep{
 				worker.Fn(StepOne).SetName("step-one").SetRateLimit(
 					worker.RateLimit{
-						Units: 1,
-						Key:   "api1",
+						Key:            "per-user-rate-limit",
+						KeyExpr:        &keyExpr,
```
Comment (grutt) [inline examples/rate-limit/main.go:85]: accepts both a key and a key expr at the same time?

File: pkg/repository/prisma/workflow.go [hunk 5]
```diff
@@ -923,16 +924,81 @@ func (r *workflowEngineRepository) createJobTx(ctx context.Context, tx pgx.Tx, t
...
 		if len(stepOpts.RateLimits) > 0 {
+			createStepExprParams := dbsqlc.CreateStepExpressionsParams{
+				Stepid: sqlchelpers.UUIDFromStr(stepId),
+			}
+
 			for _, rateLimit := range stepOpts.RateLimits {
-				_, err := r.queries.CreateStepRateLimit(
+				// if ANY of the step expressions are not nil, we create ALL options as expressions, but with static
+				// keys for any nil expressions.
+				if rateLimit.KeyExpr != nil || rateLimit.LimitExpr != nil || rateLimit.UnitsExpr != nil {
```
Comment (grutt) [inline pkg/repository/prisma/workflow.go:934]: this feels prime for missing a case, quick unit test?

File: pkg/scheduling/fixtures/sticky_soft_output.json [hunk 6]
```diff
@@ -3,10 +3,7 @@
     "16fa711e-c03e-435c-88d9-62adf1591d98",
     "064e787a-6cfd-4f82-8b6d-d8031459fdee"
   ],
-  "StepRunTimeouts": [
-    "60s",
-    "60s"
-  ],
+  "StepRunTimeouts": ["60s", "60s"],
```
Comment (grutt) [inline pkg/scheduling/fixtures/sticky_soft_output.json:6]: what happed to linting lol

File: prisma/schema.prisma [hunk 7]
```diff
@@ -830,23 +830,30 @@ model Step {
...
-
   workerLabels StepDesiredWorkerLabel[]
 
   // readable ids are unique per job
   @@unique([jobId, readableId])
 }
 
+enum StepRateLimitKind {
+  // DEFAULT rate limits are assigned to every step run
+  DEFAULT
```
Comment (grutt) [inline prisma/schema.prisma]: nit: `STATIC`?

**Review by grutt — APPROVED** [review]

File: examples/rate-limit/main.go (hunk 4 above)
Comment (abelanger5) [reply examples/rate-limit/main.go:85]: Resolved offline

**Pull request merged**